ffi/src/lib.rs

```diff
@@ -602,30 +730,50 @@ pub unsafe extern "C" fn visit_schema(
     fn visit_struct_fields(visitor: &EngineSchemaVisitor, s: &StructType) -> usize {
         let child_list_id = (visitor.make_field_list)(visitor.data, s.fields.len());
         for field in s.fields() {
-            visit_field(visitor, child_list_id, field);
+            visit_schema_item(field.data_type(), field.name(), visitor, child_list_id);
         }
         child_list_id
     }
 
+    fn visit_array_item(visitor: &EngineSchemaVisitor, at: &ArrayType) -> usize {
+        let child_list_id = (visitor.make_field_list)(visitor.data, 1);
+        visit_schema_item(&at.element_type, "array_data_type", visitor, child_list_id);
+        child_list_id
+    }
+
+    fn visit_map_types(visitor: &EngineSchemaVisitor, mt: &MapType) -> usize {
+        let child_list_id = (visitor.make_field_list)(visitor.data, 2);
+        visit_schema_item(&mt.key_type, "map_key_type", visitor, child_list_id);
+        visit_schema_item(&mt.value_type, "map_value_type", visitor, child_list_id);
+        child_list_id
+    }
+
     // Visit a struct field (recursively) and add the result to the list of siblings.
-    fn visit_field(visitor: &EngineSchemaVisitor, sibling_list_id: usize, field: &StructField) {
-        let name: &str = field.name.as_ref();
-        match &field.data_type {
-            DataType::Primitive(PrimitiveType::Integer) => {
-                (visitor.visit_integer)(visitor.data, sibling_list_id, name.into())
-            }
-            DataType::Primitive(PrimitiveType::Long) => {
-                (visitor.visit_long)(visitor.data, sibling_list_id, name.into())
-            }
-            DataType::Primitive(PrimitiveType::String) => {
-                (visitor.visit_string)(visitor.data, sibling_list_id, name.into())
-            }
-            DataType::Struct(s) => {
-                let child_list_id = visit_struct_fields(visitor, s);
-                (visitor.visit_struct)(visitor.data, sibling_list_id, name.into(), child_list_id);
-            }
-            other => println!("Unsupported data type: {}", other),
-        }
+    fn visit_schema_item(
+        data_type: &DataType,
+        name: &str,
+        visitor: &EngineSchemaVisitor,
+        sibling_list_id: usize,
+    ) {
+        // See macro def for how Struct/Map/Array/Decimal are handled
+        gen_visit_match!(
+            data_type,
+            visitor,
+            sibling_list_id,
+            name,
+            (String, visit_string),
+            (Long, visit_long),
```

Comment:
IMO this might be "too clever" for others to grok and maintain easily. And I'm not sure it really saves _that_ much vs. multiple invocations of a simpler macro for primitive types?
<details>

```rust
match data_type {
    DataType::Primitive(PrimitiveType::Decimal(precision, scale)) => {
        (visitor.visit_decimal)(visitor.data, sibling_list_id, name.into(), *precision, *scale);
    }
    DataType::Array(at) => {
        let child_list_id = visit_array_item(visitor, at);
        (visitor.visit_array)(visitor.data, sibling_list_id, name.into(), at.contains_null, child_list_id);
    }
    DataType::Struct(s) => {
        let child_list_id = visit_struct_fields(visitor, s);
        (visitor.visit_struct)(visitor.data, sibling_list_id, name.into(), child_list_id);
    }
    DataType::Map(mt) => {
        let child_list_id = visit_map_types(visitor, mt);
        (visitor.visit_map)(visitor.data, sibling_list_id, name.into(), child_list_id);
    }
    gen_visit_primitive_case!(String, visit_string, visitor, sibling_list_id, name);
      ...
    gen_visit_primitive_case!(TimestampNtz, visit_timestamp_ntz, visitor, sibling_list_id, name);
}
```

</details>

Reply:
Grr... 
```
   = note: the usage of `gen_visit_primitive_case!` is likely invalid in pattern context
   = note: macros cannot expand to match arms
```

Reply:
Ah! We can define a macro locally in the function body, which doesn't need to be hygienic:
```rust
    fn visit_schema_item(
        data_type: &DataType,
        name: &str,
        visitor: &EngineSchemaVisitor,
        sibling_list_id: usize,
    ) {
      macro_rules! visit_type {
        ( $visitor_fn:ident $(, $extra_arg:expr), *) => {
            (visitor.$visitor_fn)(visitor.data, sibling_list_id, name.into() $(, $extra_arg) *)
        };
      }
      match data_type {
          DataType::Struct(s) => {
              let child_list_id = visit_struct_fields(visitor, s);
              visit_type!(visit_struct, child_list_id);
          }
          DataType::Map(mt) => {
              let child_list_id = visit_map_types(visitor, mt);
              visit_type!(visit_map, child_list_id);
          }
          DataType::Array(at) => {
              let child_list_id = visit_array_item(visitor, at);
              visit_type!(visit_array, at.contains_null, child_list_id);
          }
          DataType::Primitive(PrimitiveType::Decimal(precision, scale)) => {
              visit_type!(visit_decimal, *precision, *scale);
          }
          DataType::STRING => visit_type!(visit_string),
            ...
          DataType::TIMESTAMP_NTZ => visit_type!(visit_timestamp_ntz),
      }
    }
```

Reply:
Aside: I wonder if `Decimal` is _really_ a `PrimitiveType`? Or should it be its own `DecimalType`?

Put another way: what benefit do we gain by having it be under the `PrimitiveType` enum? 
Seems like its args anyway make it a special case everywhere?

Reply:
Nice, the macro is much better like that!

Re `Decimal`, I agree. However, since we want to have delta-rs adopt the kernel types (see https://github.com/delta-io/delta-rs/issues/2489), and they have `Decimal` under `PrimitiveType`, I think let's leave it for now and hopefully get things adopted in delta-rs, and then we can make the change to pull it out if we still want to.